enhancements/machine-config/install-time-support-for-image-mode-on-openshift.md

```diff
@@ -0,0 +1,343 @@
+  renderedImagePushSecret:
+    name: push-secret
+```
+
+Here is an example of the "fake" MachineOSBuild generated after the cluster has been installed:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1                                 
+  kind: MachineOSBuild                                             
+  metadata:                                           
+    name: "worker-abc123"       
+    labels:                          
+      machineconfiguration.openshift.io/machineosconfig: "worker"
+      machineconfiguration.openshift.io/target-machine-config-pool: "worker"
+      # Special label marking this as synthetic       
+      machineconfiguration.openshift.io/pre-built-image: "true"        
+    annotations:         
+      machineconfiguration.openshift.io/machine-os-config: "worker"
+  spec:                        
+    desiredConfig:                       
+      name: "rendered-worker-abc123def"                 
+    machineOSConfig:                
+      name: "worker"              
+    renderedImagePushspec: "registry.example.com/custom-rhcos-worker:latest"
+  status:                        
+    buildStart: "2024-01-15T10:29:55Z"                       
+    buildEnd: "2024-01-15T10:30:00Z"                  
+    # The pre-built image is set as the build result               
+    digestedImagePushSpec: "registry.example.com/custom-rhcos-worker:latest@sha256:abc123def456..."
+```
+
+
+We will use the same MachineOSConfig and MachineOSBuild CRD definitions as already used for post-install customizations with some new annotations. The API for this can be found [here](https://github.com/openshift/api/blob/master/machineconfiguration/v1).
+
+To know more about the post-install image mode workflow, please refer to the enhancement [here](https://github.com/openshift/enhancements/pull/1515/files?short_path=9f0c5f1#diff-9f0c5f1adabad0dfbdb3c9a5b66e53d4fc6619274d7a4c260508d148de17f5c1).
+
+### Risks and Mitigations
+
+#### Failures During Installation
+
+- Implement the same failure handling as existing installation failures
+- Provide detailed guidelines on how to recover from failed installation based on the different errors
+- Provide detailed logging and error reporting for image mode build failures
+- Collect debug logs, metrics, and events for the install log bundle
+
+#### Risk: Registry Connectivity Issues
+
+- Validate registry connectivity and credentials during installation planning or bootstrap phase
+- Provide clear error messages for registry-related failures
+- Stop the install process as soon as we determine that we cannot connect to the registry
+
+## Test Plan
+
+#### Unit Tests
+
+- Bootstrap MCO MOSC parsing logic
+- Build controller bootstrap mode functionality
+- Registry integration components
+
+#### Integration Tests
+
+- Complete install time image mode on OpenShift installation workflow
+- Multiple machine config pool scenarios
+
+#### E2E Tests
```

Comment:
Testing will be pretty involved I think. Maybe we can expand on that a bit more here?

Reply:
Agreed! I will elaborate on this a bit more once I have a better idea of what it would entail